### PR TITLE
WIP: Supporting Python 2/3

### PR DESCRIPTION
**Pull request by jakirkham**
This attempts to modernize the Python codebase to support both Python 2 and Python 3. These changes were generated initially using [futurize]( http://python-future.org/automatic_conversion.html#futurize-py2-to-py2-3 ).

### CODE REVIEW

**Comment by kripken** [comment]
@saschanaz did a bunch of work on this, and we test on both python 2 and 3 currently. Not aware of any known bugs there. Does this PR fix anything concrete, or is it style stuff?

**Comment by saschanaz** [comment]
I can see so many, if not all, redundant changes...

**Comment by jakirkham** [comment]
Python 2 vs. Python 3 had many breaking changes. Am trying to help address them since I've been through this a few times unfortunately. Though the work here is incomplete. It's based largely off of a conversion tool.

One of the most notable fixes is the change of `print` from a statement to a function, which the first commit fixes. This actually caused me issues when trying to use Emscripten on Python 3.

Valid Python 2/Invalid Python 3:

```python
print "Hello World!"
```

Valid Python 2/Valid Python 3:

```python
from __future__ import print_function
print("Hello World!")
```

There are a bunch of other things that we will need to look more closely at like `unicode` vs. `bytes`. Not sure how much the encoding of strings matters for Emscripten or more importantly where it matters.

Feedback welcome.

**Comment by saschanaz** [comment]
@jakirkham Emscripten already officially supports Python 3, which is why I'm calling those changes redundant. Have you met some issue when using Emscripten with Py3?

**Comment by jakirkham** [comment]
Sorry I don't follow. Is this PR'd against the wrong branch or is there another PR open making these changes?

**Comment by saschanaz** [comment]
I'm assuming you opened this PR because you got some Py3 compat problem. It'll be helpful if you describe about the problem(s).

>is there another PR open making these changes?

There are several PRs merged: https://github.com/kripken/emscripten/pulls?utf8=%E2%9C%93&q=is:pr+python+3+compatibility+is:closed+

File: emscripten.py [hunk 1]
```diff
@@ -152,9 +158,9 @@ def fixup_metadata_tables(metadata):
   # here)
   if shared.Settings.EMULATE_FUNCTION_POINTER_CASTS and not shared.Settings.WASM:
     max_size = 0
-    for k, v in metadata['tables'].items():
+    for k, v in list(metadata['tables'].items()):
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply emscripten.py]: Agreed. Am not sure why `futurize` decides to do this.

File: tests/benchmark_sse1.py [hunk 2]
```diff
@@ -282,15 +286,15 @@ def format_comparison(a, b):
 
         }, {
             name: 'Native SSE1',
-            data: [''' + str(total_time_native_simd/total_time_native_scalar) + ''']
+            data: [''' + str(old_div(total_time_native_simd,total_time_native_scalar)) + ''']
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply tests/benchmark_sse1.py]: Yeah normally when I make this sort of change myself, I go for `//`, which is the same as C division, but works on Python 2 and Python 3 out-of-the-box.

File: third_party/websockify/tests/test_websocketproxy.py [hunk 3]
```diff
@@ -28,7 +31,7 @@
 from websockify import auth_plugins
 
 try:
-    from StringIO import StringIO
+    from io import StringIO
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply third_party/websockify/tests/test_websocketproxy.py]: Yeah this also doesn't make sense here as they already seem to have some Python 2/3 logic.

File: site/source/_themes/emscripten_sphinx_rtd_theme/__init__.py [hunk 4]
```diff
@@ -3,6 +3,7 @@
 From https://github.com/ryan-roemer/sphinx-bootstrap-theme.
 
 """
+from builtins import str
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply site/source/_themes/emscripten_sphinx_rtd_theme/__init__.py]: With future, this ends up being `from builtins import unicode as str`. So has a different effect on Python 2. That said, we might want to be cautious of this anyways.

File: emcc.py [hunk 5]
```diff
@@ -22,7 +22,16 @@
 """
 
 from __future__ import print_function
-
+from __future__ import division
+
+from future import standard_library
+standard_library.install_aliases()
```
[the remark this answers is not in the document]
Comment (jakirkham) [reply emcc.py]: We can probably revert this piece. Doubt it is needed.

**Comment by jakirkham** [comment]
Thanks for the comments @rth. Largely agree that this is an imperfect solution. Guess at this point am trying to determine whether it is worthwhile to continue pursuing this or not.

**Comment by kripken** [comment]
I agree with the overall sentiment here, my feeling is that the automatic tool is overkill. If we find specific stuff that doesn't work in Python 3, we should definitely focus on fixing that (and the automatic tool might help there, maybe).

All the main scripts should already be Python 3 compatible, at least as far as we know, but yeah, ones like `tests/benchmark_sse1.py`, that are rarely used, have not been converted. Would be nice to convert them.

**Comment by kripken** [comment]
Makes sense, updating to `print` function calls sounds useful.

**Comment by jakirkham** [comment]
Yeah I don’t have time to follow through here. Feel free to close. Though also feel free to take commits that work or make changes as you see fit.

**Comment by sbc100** [comment]
I think are python3 support is pretty good today. We run all the tests under python3.

**Comment by jakirkham** [comment]
Ah ok. That’s great to hear! Thanks for the update 😀

**Pull request closed without merging**